src/main/java/org/folio/service/UserService.java

```diff
@@ -0,0 +1,161 @@
+package org.folio.service;
+
+import io.vertx.core.Future;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.client.HttpRequest;
+import io.vertx.ext.web.client.HttpResponse;
+import io.vertx.ext.web.client.WebClient;
+import io.vertx.ext.web.client.predicate.ResponsePredicate;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.folio.config.model.SamlConfiguration;
+import org.folio.okapi.common.XOkapiHeaders;
+import org.folio.util.StringUtil;
+import org.folio.util.model.OkapiHeaders;
+import org.pac4j.core.context.session.SessionStore;
+import org.pac4j.core.profile.UserProfile;
+import org.pac4j.saml.client.SAML2Client;
+import org.pac4j.saml.credentials.SAML2Credentials;
+import org.pac4j.vertx.VertxWebContext;
+
+import javax.ws.rs.core.UriBuilder;
+import java.util.Collections;
+import java.util.List;
+
+import static org.folio.util.UserFields.*;
+
+public class UserService {
+
+  private static final Logger log = LogManager.getLogger(UserService.class);
+
+  private static final String USER_TENANT_ENDPOINT = "/user-tenants";
+  private static final String USER_TENANT_GET_ERROR = "Error getting user-tenant by %s = %s: %s";
+  private static final String MULTIPLE_MATCHING_USERS = "Multiple matching users by {} = {}";
+  private static final String USER_TENANT_DOES_NOT_HAVE_A_TENANT_ID_ERROR = "The matching user-tenant record founded by %s = %s does not have a tenant id";
+
+  public static class UserErrorException extends RuntimeException {
+    public UserErrorException(String message) {
+      super(message);
+    }
+  }
+
+  public static class UserFetchException extends RuntimeException {
+    public UserFetchException(String message) {
+      super(message);
+    }
+  }
+
+  public Future<JsonObject> getUser(WebClient webClient, SamlConfiguration configuration,
+                                              VertxWebContext webContext, SAML2Client client, SessionStore sessionStore,
+                                              OkapiHeaders parsedHeaders) {
+    final String userPropertyName =
+      configuration.getUserProperty() == null ? EXTERNAL_SYSTEM_ID : configuration.getUserProperty();
+    var credentialsOptional = client.getCredentials(webContext, sessionStore);
+    var credentials =
+      (SAML2Credentials) credentialsOptional.orElseThrow(() -> new NullPointerException("Saml credentials was null"));
+
+    final String samlAttributeValue =
```

Comment:
**Minor**
I'd remove all `final`s according to "use everywhere or nowhere" rule.

Reply:
Removed